pennylane/pauli/conversion.py

```diff
@@ -19,36 +19,52 @@
 from operator import matmul
 from typing import Union
 
-import numpy as np
-
+import pennylane as qml
 from pennylane.operation import Tensor
 from pennylane.ops import Hamiltonian, Identity, PauliX, PauliY, PauliZ, Prod, SProd, Sum
+from pennylane.ops.qubit.matrix_ops import _walsh_hadamard_transform
 
-from .pauli_arithmetic import I, PauliSentence, PauliWord, X, Y, Z, mat_map, op_map
+from .pauli_arithmetic import I, PauliSentence, PauliWord, X, Y, Z, op_map
 from .utils import is_pauli_word
 
 
+# pylint: disable=too-many-branches
 def pauli_decompose(
-    H, hide_identity=False, wire_order=None, pauli=False
+    matrix,
+    hide_identity=False,
+    wire_order=None,
+    pauli=False,
+    padding=False,
 ) -> Union[Hamiltonian, PauliSentence]:
-    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+    r"""Decomposes a matrix into a linear combination of Pauli operators.
+
+    This method converts any matrix to a weighted sum of Pauli words acting on :math:`n` qubits
+    in time :math:`O(n 4^n)`. The input matrix is first padded with zeros if its dimensions are not
+    :math:`2^n\times 2^n` and written as a quantum state in the computational basis following the
+    `channel-state duality <https://en.wikipedia.org/wiki/Channel-state_duality>`_.
+    A Bell basis transformation is then performed using the
+    `Walsh-Hadamard transform <https://en.wikipedia.org/wiki/Hadamard_transform>`_, after which
+    coefficients for each of the :math:`4^n` Pauli words are computed while accounting for the
+    phase from each ``PauliY`` term occurring in the word.
 
     Args:
-        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        matrix (tensor_like[complex]): any matrix M, the keyword argument ``padding=True``
+            should be provided if the dimension of M is not :math:`2^n\times 2^n`.
         hide_identity (bool): does not include the Identity observable within
             the tensor products of the decomposition if ``True``.
         wire_order (list[Union[int, str]]): the ordered list of wires with respect
             to which the operator is represented as a matrix.
         pauli (bool): return a PauliSentence instance if ``True``.
+        padding (bool): makes the function compatible with rectangular matrices and square matrices
+            that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
```

Comment:
> Can we decompose a rectangle matrix without padding?
No, we cannot. 

I prefer keeping a `kwarg` to prevent scenarios where users mistakenly input a matrix of a non-standard shape. They would not know they are doing so unless we raise a `warning` in that case. I want them to be aware of what they are actually doing.